browser/brave_wallet/fil_nonce_tracker_unittest.cc

```diff
@@ -0,0 +1,208 @@
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "chrome/browser/prefs/browser_prefs.h"
+#include "chrome/test/base/testing_profile_manager.h"
+#include "components/sync_preferences/testing_pref_service_syncable.h"
+#include "content/public/test/browser_task_environment.h"
+#include "services/network/public/cpp/weak_wrapper_shared_url_loader_factory.h"
+#include "services/network/test/test_url_loader_factory.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+class FilNonceTrackerUnitTest : public testing::Test {
+ public:
+  FilNonceTrackerUnitTest() {
+    shared_url_loader_factory_ =
+        base::MakeRefCounted<network::WeakWrapperSharedURLLoaderFactory>(
+            &url_loader_factory_);
+  }
+
+  void SetUp() override {
+    TestingProfile::Builder builder;
+    auto prefs =
+        std::make_unique<sync_preferences::TestingPrefServiceSyncable>();
+    RegisterUserProfilePrefs(prefs->registry());
+    builder.SetPrefService(std::move(prefs));
+    profile_ = builder.Build();
+  }
+
+  PrefService* GetPrefs() { return profile_->GetPrefs(); }
+
+  network::SharedURLLoaderFactory* shared_url_loader_factory() {
+    return url_loader_factory_.GetSafeWeakWrapper().get();
+  }
+
+  void WaitForResponse() { task_environment_.RunUntilIdle(); }
+
+  void SetTransactionCount(uint64_t count) {
+    transaction_count_ = count;
+    url_loader_factory_.ClearResponses();
+
+    // See JsonRpcService::SetNetwork() to better understand where the
+    // http://localhost:7545 URL used below is coming from.
+    url_loader_factory_.AddResponse(
+        brave_wallet::GetNetworkURL(GetPrefs(), mojom::kLocalhostChainId,
+                                    mojom::CoinType::FIL)
+            .spec(),
+        GetResultString());
+  }
+
+ private:
+  std::string GetResultString() const {
+    return "{\"id\":1,\"jsonrpc\":\"2.0\",\"result\":" +
+           base::NumberToString(transaction_count_) + "}";
+  }
+
+  uint64_t transaction_count_ = 0;
+  network::TestURLLoaderFactory url_loader_factory_;
+  scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory_;
+  content::BrowserTaskEnvironment task_environment_;
+  std::unique_ptr<TestingProfile> profile_;
+};
+
+TEST_F(FilNonceTrackerUnitTest, GetNonce) {
+  JsonRpcService service(shared_url_loader_factory(), GetPrefs());
+  base::RunLoop run_loop;
+  service.SetNetwork(
+      brave_wallet::mojom::kLocalhostChainId, mojom::CoinType::FIL,
+      base::BindLambdaForTesting([&](bool success) { run_loop.Quit(); }));
+  run_loop.Run();
+
+  FilTxStateManager tx_state_manager(GetPrefs(), &service);
+  FilNonceTracker nonce_tracker(&tx_state_manager, &service);
+
+  SetTransactionCount(2);
+
+  const std::string addr("t1lqarsh4nkg545ilaoqdsbtj4uofplt6sto26ziy");
+  uint64_t nonce_result = 0;
+  bool callback_called = false;
+  // tx count: 2, confirmed: null, pending: null
+  nonce_tracker.GetNextNonce(
```

Comment:
Suggest to make it a member function in FilNonceTrackerUnitTest and use run_loop.Quit() in the lambda.

Reply:
Can pass in expected nonce into that member function and check inside the lambda too.

Reply:
done